app/controllers/users/nbp_wallet_controller.rb

```diff
@@ -0,0 +1,82 @@
+# frozen_string_literal: true
+
+require 'rqrcode'
+
+module Users
+  class NbpWalletController < ApplicationController
+    skip_after_action :verify_authorized
+
+    def connect
+      if Enmeshed::ConnectorApi.pending_relationship_for_nbp_uid(@provider_uid).present?
+        redirect_to nbp_wallet_finalize_users_path and return
+      end
+
+      @truncated_reference = Enmeshed::ConnectorApi.create_relationship_template(@provider_uid)
+
+      @template_url = "nmshd://tr##{@truncated_reference}"
+    rescue Enmeshed::ConnectorError, Faraday::Error => e
+      Sentry.capture_exception(e)
+      redirect_to new_user_registration_path, alert: t('common.errors.generic_try_later')
+    end
+
+    def qr_code
+      send_data RQRCode::QRCode.new("nmshd://tr##{params[:truncated_reference]}").as_png(border_modules: 0), type: 'image/png'
+    end
+
+    def relationship_status
+      if Enmeshed::ConnectorApi.pending_relationship_for_nbp_uid(@provider_uid).present?
+        render json: {status: :ready}
+      else
+        render json: {status: :waiting}
+      end
+    rescue Enmeshed::ConnectorError, Faraday::Error => e
+      Sentry.capture_exception(e)
+      redirect_to nbp_wallet_connect_users_path, alert: t('common.errors.generic')
+    end
+
+    def finalize
+      @relationship = Enmeshed::ConnectorApi.pending_relationship_for_nbp_uid(@provider_uid)
+      abort_and_refresh and return if @relationship.blank?
+
+      accept_and_create_user
+    rescue Enmeshed::ConnectorError, Faraday::Error => e
+      Sentry.capture_exception(e)
+      abort_and_refresh
+    end
+
+    private
+
+    def accept_and_create_user # rubocop:disable Metrics/AbcSize
+      @user = User.new_from_omniauth(@relationship.userdata, 'nbp', @provider_uid)
+      @user.identities << UserIdentity.new(omniauth_provider: 'enmeshed', provider_uid: @relationship.json[:peer])
+      @user.skip_confirmation_notification!
+
+      ApplicationRecord.transaction do
+        unless @user.save
+          abort_and_refresh(t('common.errors.model_not_created', model: User.model_name.human,
+            errors: @user.errors.full_messages.join(', ')))
+          raise ActiveRecord::Rollback
+        end
+
+        if @relationship.accept
+          @user.send_confirmation_instructions
+          redirect_to home_index_path, notice: t('devise.registrations.signed_up_but_unconfirmed')
+        else
+          abort_and_refresh
+          raise ActiveRecord::Rollback
+        end
+      end
+    end
+
+    def abort_and_refresh(reason = t('common.errors.generic'))
+      Rails.logger.debug { "NbpWalletController calling abort_and_refresh() due to: #{reason}" }
+      @relationship&.reject
+      redirect_to nbp_wallet_connect_users_path, alert: reason
+    end
```

Comment:
Rejecting the template could still fail, so we should handle this error, too:

```suggestion
    def abort_and_refresh(reason = t('common.errors.generic'))
      Rails.logger.debug { "NbpWalletController calling abort_and_refresh() due to: #{reason}" }
      @relationship&.reject
    rescue Enmeshed::ConnectorError, Faraday::Error => e
      # We still need to capture and handle an HTTP error caused by rejecting the template.
      Sentry.capture_exception(e)
    finalize
      redirect_to nbp_wallet_connect_users_path, alert: reason
    end
```

Reply:
Done.